### PR TITLE
ARROW-2286: [C++/Python] Allow subscripting pyarrow.lib.StructValue

### PR DESCRIPTION
**Pull request by kszucs**

### CODE REVIEW

**Comment by pitrou** [comment]
Thanks @kszucs. Can you rebase to fix the CI failures?

**Comment by kszucs** [comment]
@pitrou rebase done, but the builds are still failing

**Comment by pitrou** [comment]
Ok, so the Travis-CI failure is unrelated (we are getting npm errors intermittently: see ARROW-2509).

However, the Windows compile errors are probably caused by this PR. I'll point out the issue in a comment.

File: python/pyarrow/scalar.pxi [hunk 1]
```diff
@@ -371,14 +371,29 @@ cdef class FixedSizeBinaryValue(ArrayValue):
...
+            CStructType* type
+            int64_t index
+
+        type = <CStructType*> self.type.type
+        index = type.GetChildIndex(tobytes(key))
+
+        if index < 0:
+            raise KeyError(key)
+
+        return pyarrow_wrap_array(self.ap.field(index))[self.index]
```
Comment (pitrou) [inline python/pyarrow/scalar.pxi]: You are passing an int64_t to a function that takes a (32-bit) int, hence the warning about truncation which is turned into an error.

File: cpp/src/arrow/type.h [hunk 2]
```diff
@@ -475,6 +475,15 @@ class ARROW_EXPORT StructType : public NestedType {
   Status Accept(TypeVisitor* visitor) const override;
   std::string ToString() const override;
   std::string name() const override { return "struct"; }
+
+  /// Returns null if name not found
+  std::shared_ptr<Field> GetChildByName(const std::string& name) const;
+
+  /// Returns -1 if name not found
+  int64_t GetChildIndex(const std::string& name) const;
```
Comment (pitrou) [inline cpp/src/arrow/type.h]: Child indices are `int` everywhere else, not `int64_t`, so this function should return `int` as well.

File: cpp/src/arrow/type.h [hunk 3]
```diff
@@ -475,6 +475,15 @@ class ARROW_EXPORT StructType : public NestedType {
...
+
+ private:
+  mutable std::unordered_map<std::string, int> name_to_index_;
```
Comment (pitrou) [inline cpp/src/arrow/type.h]: Add a comment that this is lazily initialized?

File: cpp/src/arrow/type-test.cc [hunk 4]
```diff
@@ -399,6 +399,40 @@ TEST(TestStructType, Basics) {
...
+  auto f0 = field("f0", int32());
+  auto f1 = field("f1", uint8(), false);
+  auto f2 = field("f2", utf8());
+  auto f3 = field("f3", list(int16()));
+
+  StructType struct_type({f0, f1, f2, f3});
+  std::shared_ptr<Field> result;
+
+  result = struct_type.GetChildByName("f1");
+  ASSERT_TRUE(f1->Equals(result));
```
Comment (pitrou) [inline cpp/src/arrow/type-test.cc]: You could actually check for pointer equality here, i.e. `ASSERT_EQ(f1, result)`.

File: cpp/src/arrow/type-test.cc [hunk 5]
```diff
@@ -399,6 +399,40 @@ TEST(TestStructType, Basics) {
...
+
+  result = struct_type.GetChildByName("f3");
+  ASSERT_TRUE(f3->Equals(result));
+
+  result = struct_type.GetChildByName("not-found");
+  ASSERT_TRUE(result == nullptr);
```
Comment (pitrou) [inline cpp/src/arrow/type-test.cc]: `ASSERT_EQ` will produce nicer diagnostics on failure.

File: python/pyarrow/tests/test_scalars.py [hunk 6]
```diff
@@ -215,3 +215,21 @@ def test_array_to_set(self):
         set_from_array = set(arr)
         assert isinstance(set_from_array, set)
         assert set_from_array == {1, 2}
+
+    def test_struct_array_subscripting(self):
```
Comment (pitrou) [inline python/pyarrow/tests/test_scalars.py]: Nit: call this `test_struct_value_subscripting`? You're not subscripting the array here.

**Review by pitrou — CHANGES_REQUESTED** [review]
See my comments below about the likely cause of the Windows failure (and some more minor comments).

**Review by pitrou — APPROVED** [review]
Thanks @kszucs for doing this! LGTM now.

**Pull request closed without merging**